File: percy+ann/percy.hs
import Matrix
import System.Random

{- | The output follows the high-bit of the input.  We want the perceptron
 --  percieve this. -}
training_inputs=[[0, 0, 1], [1, 1, 1], [1, 0, 1], [0, 1, 1]] -- ::[[Float]]
training_outputs=transpose [[0, 1, 1, 0]]::[[Float]]

sigmoid x =  1 / (1 + exp (-x))
sigderiv sx = sx * (1 - sx)

-- | Calculate a new output and weights ("think" function)
out_wout inp win = let out = mmap sigmoid (inp `mmult` win)
                       err = matsub training_outputs out
                       adj = err `hmul` (mmap sigderiv out)
                       wout= madd win ((transpose inp) `mmult` adj)
                    in (out, wout) 

ann :: F -> F -> F -> Int -> (F,F)
ann  trn_in weights trn_out iter =
  let ann' :: F -> F -> Int -> (F,F)
      ann' o w i = if i <= 0 then (o,w) else let (o',w') = out_wout training_inputs w in ann' o' w'  (i - 1)
      (out,weights') = out_wout training_inputs weights
   in ann' out weights' (iter -1)
                   
main = 
  do rx <- getStdGen >>= return . randoms >>= return . take 3 :: IO [Float]
     let synaptic_weights = transpose [ map (\r -> (2 * r) - 1) rx ]
     
     putStrLn $ "Training inputs:\n" ++ show training_inputs ++ "\n\nTraining outputs:\n" ++show training_outputs
     
     putStrLn $ "\nRandom starting synaptic weights:\n" ++ show synaptic_weights
     
     let (out,ws_out) = ann training_inputs synaptic_weights training_outputs 20000
     putStrLn $ "\nWeights after training: " ++ show ws_out ++ "\n\nOutput  after training: " ++ show out

     let (newCase,(newOutput,_)) = ([[1,0,0]], out_wout newCase ws_out)
     putStrLn $ "\nNew Input and Output with trained weights:" ++ show  newCase ++ "=>" ++ show newOutput 

File: percy+ann/percy.py
import numpy as np

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

# "*" is the Hadamard product
def sigmoid_derivative(sx):
    return sx * (1 - sx)

training_inputs = np.array([[0,0,1],
                            [1,1,1],
                            [1,0,1],
                            [0,1,1]])

training_outputs = np.array([[0,1,1,0]]).T

"""np.random.seed(1) # Not used"""

synaptic_weights = 2 *  np.random.random((3,1)) - 1

print("Training inputs:",  training_inputs)
print("Training outputs:", training_outputs)
print("Random starting synaptic weights:", synaptic_weights)

input_layer = training_inputs

for iteration in range(20000):

    # np.dot is regular matrix* in this case
    output = sigmoid(np.dot(input_layer,synaptic_weights))

    error = training_outputs - output
    # '*' is the Hadamard product
    adjustments = error * sigmoid_derivative(output)
    #print("*\nerror\n", error, "\nderiv\n", adjustments, "\n\n")

    synaptic_weights += np.dot(input_layer.T,  adjustments)
    

print("Synaptic weights after training:", synaptic_weights)

print("Outputs after training", output)

novelout = sigmoid(np.dot([1,0,0],synaptic_weights))

print("Output for novel input: [ 1,0,0 ] => ", novelout)
